### PR TITLE
Add slice parameter type checking to disallow Tensor usage for slices 

### PR DESCRIPTION
**Pull request by mnovosad1095**
Added behaviour matching PyTorch when using tensors for slicing 
Pls correct me if we are not supposed to support single element tensors when slicing
#6900

### CODE REVIEW

File: test/imported/test_indexing.py [hunk 1]
```diff
@@ -98,6 +98,9 @@ def test_index(self):
     numpy_testing_assert_equal_helper(reference[2, 2, 2], 27)
     numpy_testing_assert_equal_helper(reference[:], consec((3, 3, 3)))
 
+    # indexing with single element integer tensor
```
Comment (chenyuxyz) [inline test/imported/test_indexing.py:101]: don't add test directly to this file, it's a copy of torch's test.

add these tests to test_ops close to other fancy getitem tests

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -1023,11 +1023,28 @@ def pad(self, arg:Tuple[Optional[Tuple[sint, sint]], ...], value:float=0.0) -> T
   #   3. Out of bounds Tensor indexing results in 0
   #     - e.g: Tensor([1, 2, 3])[Tensor([4, 3, 2])] -> [0, 0, 3] index 4 and 3 are out of bounds
   def _getitem(self, indices, v: Optional[Tensor] = None) -> Tensor:
+    def _fix_slice(x: slice) -> slice:
+      # convert slice scalar Tensor slice parameters to int
+      def process_component(component):
+        if isinstance(component, Tensor):
+          if component.numel() > 1 or not dtypes.is_int(component.dtype): 
+            raise TypeError("only integer tensors of single element can be converted to a slice index")
+          return component.item()
```
Comment (chenyuxyz) [inline tinygrad/tensor.py:1032]: it should not have a `.item` here, as it becomes data dependent

Comment (mnovosad1095) [reply tinygrad/tensor.py:1032]: Got it
But then the question is, should tinygrad support single-element tensors in slices, which will involve some more clever refactoring, or not and just throw an type err that this operation is not supported?

File: test/imported/test_indexing.py (hunk 1 above)
Comment (mnovosad1095) [reply test/imported/test_indexing.py:101]: Got it, will fix

**Comment by chenyuxyz** [comment]
too many lines. please read the existing code and put the check in an appropriate place, not just adding plenty of code just for your specific usage. it should not check specifically for tinygrad.Tensor and should allow only the supported slice.

**Comment by mnovosad1095** [comment]
Thanks for your feedback! 
I've moved the check under the slice 0-step check and made it more concise and general, as you asked. 
Now the error message would look something like this:

`TypeError: Unsupported slice for dimension 0. Expected integers or None, got slice(float, int, NoneType).`

**Comment by chenyuxyz** [comment]
i think this is fine as we clearly don't support it now, would like to know how other framework handles it and we should look into matching those

**Pull request merged**